### PR TITLE
Overcommit VCPU for Scheduling

### PR DESCRIPTION
**Pull request by so-sahu**
Fixes https://github.com/ironcore-dev/libvirt-provider/issues/167

### CODE REVIEW

File: pkg/resources/sources/cpu.go [hunk 1]
```diff
@@ -31,13 +37,16 @@ func (c *CPU) GetTotalResources(ctx context.Context) (core.ResourceList, error)
...
-	resources := core.ResourceList{core.ResourceCPU: *resource.NewScaledQuantity(hostCPUSum, resource.Kilo)}
+	cpuQuantity := float64(hostCPUSum) * c.OvercommitVCPU
+	resources := core.ResourceList{
+		core.ResourceCPU: *resource.NewQuantity(int64(cpuQuantity*1000), resource.DecimalSI),
+	}
 
 	return resources, nil
 }
 
-func (c *CPU) GetName() string {
```
Comment (lukas016) [inline pkg/resources/sources/cpu.go:39]: why are you mixing source name with resource name?

File: pkg/resources/sources/cpu.go [hunk 2]
```diff
@@ -13,13 +14,18 @@ import (
...
+}
 
-func NewSourceCPU() *CPU {
-	return &CPU{}
+func NewSourceCPU(options Options) *CPU {
+	return &CPU{OvercommitVCPU: options.OvercommitVCPU}
 }
 
 func (c *CPU) GetTotalResources(ctx context.Context) (core.ResourceList, error) {
+	if c.OvercommitVCPU == 0 || c.OvercommitVCPU < 0 {
```
Comment (lukas016) [inline pkg/resources/sources/cpu.go:26]: i recommend use https://pkg.go.dev/math#Signbit instead of < 0

File: pkg/resources/sources/cpu.go (hunk 1 above)
Comment (so-sahu) [reply pkg/resources/sources/cpu.go:39]: ok, I will create separate constants for the source names.

Comment (so-sahu) [reply pkg/resources/sources/cpu.go:39]: However, we already have the source name associated with the resource name: [memory.go](https://github.com/ironcore-dev/libvirt-provider/blob/b1d44ebf879bfb8ea48a54a84198d8e827376e19/pkg/resources/sources/memory.go#L17).

Comment (lukas016) [reply pkg/resources/sources/cpu.go:39]: no, it is resource name for new resource, but in GetName() i returned hugepages strings

File: pkg/resources/sources/cpu.go [hunk 3]
```diff
@@ -31,13 +37,16 @@ func (c *CPU) GetTotalResources(ctx context.Context) (core.ResourceList, error)
 		hostCPUSum += int64(v.Cores)
 	}
 
-	resources := core.ResourceList{core.ResourceCPU: *resource.NewScaledQuantity(hostCPUSum, resource.Kilo)}
+	cpuQuantity := float64(hostCPUSum) * c.OvercommitVCPU
+	resources := core.ResourceList{
```
Comment (lukas016) [inline pkg/resources/sources/cpu.go:41]: why did you change this operation and you didn't used NewScaledQuantity?

Comment (so-sahu) [reply pkg/resources/sources/cpu.go:41]: The change was made because NewScaledQuantity doesn't handle floating-point scaling well. NewScaledQuantity expects an integer as the first argument and a resource.Scale as the second argument. Since c.OvercommitVCPU can be a float, we needed to perform the scaling manually by converting the result to an integer. This manual scaling ensures that fractional values like 1.5 or 0.5 are correctly converted to their corresponding values in kilo. Using NewScaledQuantity directly wouldn't have allowed us to handle such fractional values correctly.

File: pkg/resources/sources/cpu.go (hunk 1 above)
Comment (so-sahu) [reply pkg/resources/sources/cpu.go:39]: Got it. I'll create a separate constant for the source name and use that instead of hardcoding strings in multiple places to ensure it can be managed from a single location.

File: pkg/resources/sources/cpu.go (hunk 2 above)
Comment (lukas016) [reply pkg/resources/sources/cpu.go:26]: i don't think it will safe too. comparing float to specific number isn't (minimally in c language) safe operation.
so maybe something as c.OvercommitVCPU < 0.0001 can be safer.

File: pkg/resources/sources/cpu.go (hunk 3 above)
Comment (lukas016) [reply pkg/resources/sources/cpu.go:41]: but you don't use float and you retype it to int64, so i am confuse.

Comment (so-sahu) [reply pkg/resources/sources/cpu.go:41]: The reason for manually scaling the CPU quantity and converting it to an integer (int64) is to ensure that fractional values, such as 1.5 or 0.5, are correctly converted to their corresponding values in kilo. Even though the c.OvercommitVCPU value may be a float, the final result needs to be an integer representing the CPU quantity in kilo.

The NewScaledQuantity function expects an integer as its first argument, which is why we need to perform the scaling manually. By converting the result to an integer, we ensure that the CPU quantity is correctly represented in kilo, even if the c.OvercommitVCPU value is a float.

Using NewScaledQuantity directly wouldn't allow us to handle such fractional values correctly, as it expects an integer value for scaling. Therefore, the manual scaling and conversion to an integer are necessary in this case.

File: pkg/resources/sources/cpu.go (hunk 2 above)
Comment (so-sahu) [reply pkg/resources/sources/cpu.go:26]: Using `epsilon := 1e-9` for the comparison. It will solve the issue.

File: pkg/resources/sources/cpu.go (hunk 3 above)
Comment (so-sahu) [reply pkg/resources/sources/cpu.go:45]: As discussed offline, using `NewScaledQuantity`

**Review by lukas016 — APPROVED** [review]
